### ISSUE TITLE
Feature request: Make it a compile-time error to mark a class `@Serializable` when it has potentially non-serializable interfaces as members.

### ISSUE THREAD

**Issue by Sintrastes** [Kotlin/kotlinx.serialization#1737]
**What is your use-case and why do you need this feature?**

One of the reasons I was initially so excited to try kotlinx.serialization over existing solutions (such as GSON), was because it is a compiler plugin, so it has the capability to produce a compile-time error when trying to serialize a class that is not serializable. 

This seems to work well in most circumstances, but one area I've found where this fails (that seems like low-hanging fruit!) is in a circumstance like the following:

```kotlin
interface MyInterface

@Serializable
data class MyData(val refToInterface: MyInterface)
```

This compiles fine,  but if I try to serialize `MyData` with an instance of `MyInterface` as a member -- at runtime I get an exception saying that there are no polymorphic serializers registered.

**Describe the solution you'd like**

The compiler should either warn that the parameter should either be marked `@Contextual`, or a specific serializer needs to be specified for the parameter by `@Serializable(with = ...)`.

**Comment by Sintrastes** [comment]
> Only interfaces are automatically polymorphic, but none of their implementers.

Why would interfaces be automatically polymorphic if this just means they will crash at runtime when attempted to be serialized, since no serialization implementation is provided by default? I think that's the root of what I'm getting after.

Interfaces should not be automatically serializable via the `PolymorphicSerialization` strategy. If the user wants to use that strategy, they should explicitly have to opt-in with annotating their interface with `@SerializableWith(PolymorphicSerializer::class)`.

I can open another issue for this (as a feature request) as appropriate, as I see this has been changed to a question.

**Comment by Sintrastes** [comment]
Although, I think if I understand it correctly, [this](https://github.com/Kotlin/kotlinx.serialization/issues/1247) proposal will give me what I want.

**Comment by Sintrastes** [comment]
> Well, if you add an interface to a serializable class, there's no other way you can serialize it except polymorphically.

I don't understand this remark. See, for instance an example in Scala 3:

https://gist.github.com/Sintrastes/b8a6b0317fe82661d59ba22131a08f25

>  Open polymorphism is a powerful mechanism that is tied to the runtime by its nature, so giving up on compile-time errors is a tradeoff for convenience.

That's a fine tradeoff if some people want to be able to make it, but as I've shown in my example above, it is not a _nescesary aspect_ of using subtype polymorphism. You can use subtype polymorphism and still go about perfectly well having every serializer resolved at compile-time.  

> Thus, @Polymorphic annotation on every property with interface type is just a boilerplate.

It's only boilerplate if it's the behavior you want. I want to at least have the option of being able to opt-out, and have `@Serializable` behave exactly like a typeclass, like in Scala, which again, is perfectly possible to do in a language with subtype polymorphism.

I'd be happy to discuss more specifics of my particular use case, and why I want this feature, but I do not understand your initial remark about adding an interface to a class making it impossible to serialize any way but polymorphically.

Perhaps another property to think about is something like:

    for all A, A: B and x: A where A and B are serializable: serialize(x) == serialize(x as B)
    
In other words, if in polymorphic serialization, no information is lost via casting. If this is something that's desirable, then this would definitely _not_ hold under my proposal.

**Comment by sandwwraith** [comment]
Maybe I've used incorrect phrasing, I meant that if you have `@Serializable class A` and `interface I`, if you add a new property to A: `@Serializable class A(val i: I)`, you have to serialize this property polymorphically — determine a runtime type of `i`. Even if type discriminator is not added to the output (and you get the same output as with concrete type), this is still a polymorphic mechanism.

**Comment by sandwwraith** [comment]
Btw, if you want to hold `deserialize(serialize(x)) == x` you need to add type information to the output.

**Comment by arkivanov** [comment]
@sandwwraith This has been a significant problem in our projects, human errors are slipping to production from time to time causing the app to crash. This is the latest case:

```kotlin
@Serializable
data class MyType(
    val items: ImmutableList<String>,
)
```

We have our own `class ImmutableListSerializer<T>` and developers often forget to annotate properties with `@Serialiazable(ImmutableListSerializer::class)`. It would be much safer for us to have a way of failing compilation in such cases.

**Comment by pdvrieze** [comment]
What about having an option for the plugin that would only allow specified open polymorphic base types? As in, the plugin configuration would allow specifying a list of types, and would fail on polymorphic types that are not in the list (sealed interfaces/classes would be fine). Given that sealed interfaces are now well supported this would probably allow sufficient checks, be easy to implement, and be easy enough to work with.

**Comment by sandwwraith** [comment]
I think this can be solved with some plugin option. Or a special `@NonPolymorphic` annotation.

**Comment by pdvrieze** [comment]
@sandwwraith I think that for this case `@NonPolymorphic` might be insufficient, but both could be a good idea.
One question to consider is whether these mechanisms would be transitive. Probably it can't be transitive across module boundaries.
